lib/utility/include/Logger.h

```diff
@@ -1,28 +1,41 @@
 #pragma once
 
+#include <iostream>
+#include <mutex>
+#include <sstream>
+
+#include "FeatureFlags.h"
+
 namespace Utility {
 
-class Logger {
+class Logger;
+extern Logger logger;
+
+class Logger : public FileHandle {
  public:
-  // C-style printf
-  static int printf(const char* format, ...);
-
-  Logger& operator<<(long n);
-  Logger& operator<<(unsigned long n);
-  Logger& operator<<(bool n);
-  Logger& operator<<(short n);
-  Logger& operator<<(unsigned short n);
-  Logger& operator<<(int n);
-  Logger& operator<<(unsigned int n);
-  Logger& operator<<(long long n);
-  Logger& operator<<(unsigned long long n);
-
-  Logger& operator<<(double f);
-  Logger& operator<<(float f);
-  Logger& operator<<(long double f);
-
-  Logger& operator<<(char c);
-  Logger& operator<<(const char* format);
+  Logger();
+
+  template <typename T>
+  Logger& operator<<(T out) {
+    std::scoped_lock<Mutex> lock(m_mutex);
+
+    std::stringstream ss;
+    ss << out;
```

Comment:
There's probably a way to detect ISR context and in those situations defer execution to non-ISR context (using mbed events or similar). Or maybe just always defer execution so we minimize logger's impact on code performance?

Either way, I think it can be slated as improvements in a future ticket

Reply:
will add this to the backlog

Reply:
our logger aside, you cant call mbed's printf from isr either. i think this is low priority. " This is because calls to stdio (like printf) are guarded by mutexes in the ARM C standard library"

Reply:
Yeah that's why I said defer execution